### PR TITLE
[WIP] ✨ KCP remediation

### LINKED ISSUE #2976: KCP unhealthy remediation support
**User Story**

As an operator I would like to have my control plane clusters automatically remediate unhealthy machines.

**Detailed Description**

- The cluster should scale up a replacement machine and then scale down the unhealthy machine. Rinse and repeat.
- The cluster must contain at least one healthy control plane machine for any remediation to happen (clusters where all the nodes are unhealthy are in the best case difficult to remediate, at worst impossible)
- Clusters must have etcd quorum in order to do automatic remediation

**Anything else you would like to add:**

**TBD**: how do we stop remediation from happening in an infinite loop?

We want to move the calls to `reconcileHealth` from scale up and scale down to make it possible for them to operate on an unhealthy cluster. `reconcileHealth` should be earlier in the reconciliation loop. Right now it would (probably?) block remediation if we tried to call scaleUp/scaleDown since the cluster would not be healthy.

/kind feature

### PR DESCRIPTION
**Pull request by benmoss**
**What this PR does / why we need it**:
Adds support for remediation to KCP

**Which issue(s) this PR fixes**:
Fixes #2976

### CODE REVIEW

File: controlplane/kubeadm/controllers/controller.go [hunk 1]
```diff
@@ -232,6 +232,18 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 	}
 
 	controlPlane := internal.NewControlPlane(cluster, kcp, ownedMachines)
+
+	if requeue := preflightChecks(controlPlane); requeue != nil {
+		return *requeue, nil
+	}
+
+	if result, err := r.reconcileHealth(ctx, cluster, kcp, controlPlane); err != nil {
```
Comment (fabriziopandini) [inline controlplane/kubeadm/controllers/controller.go:240]: this is kind of hard to read considering err and the optional result, but I don't have better ideas...

File: controlplane/kubeadm/controllers/controller.go [hunk 2]
```diff
@@ -351,15 +363,27 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
...
+		return &result, err
+	}
 	// Do a health check of the Control Plane components
 	if err := r.managementCluster.TargetClusterControlPlaneIsHealthy(ctx, util.ObjectKey(cluster), kcp.Name); err != nil {
-		logger.V(2).Info("Waiting for control plane to pass control plane health check to continue reconciliation", "cause", err)
+		logger.Info("Waiting for control plane to pass control plane health check to continue reconciliation", "cause", err)
 		r.recorder.Eventf(kcp, corev1.EventTypeWarning, "ControlPlaneUnhealthy",
 			"Waiting for control plane to pass control plane health check to continue reconciliation: %v", err)
-		return &capierrors.RequeueAfterError{RequeueAfter: healthCheckFailedRequeueAfter}
+		return &ctrl.Result{RequeueAfter: healthCheckFailedRequeueAfter}, nil
```
[the remark this answers is not in the document]
Comment (benmoss) [reply controlplane/kubeadm/controllers/controller.go:386]: Hmm, I hadn't thought through `RequeueAfterError` entirely, I just wanted to change this to distinguish errors from requeues. It was frustrating when trying to debug this that the error being returned was just a `RequeueAfterError` with no meaningful information about what error happened. Maybe if we used `errors.Wrap(err, RequeueAfterError{})` instead that'd be better?

File: controlplane/kubeadm/controllers/controller.go [hunk 3]
```diff
@@ -368,23 +392,38 @@ func (r *KubeadmControlPlaneReconciler) reconcileHealth(ctx context.Context, clu
...
 	}
 
-	// We need this check for scale up as well as down to avoid scaling up when there is a machine being deleted.
-	// This should be at the end of this method as no need to wait for machine to be completely deleted to reconcile etcd.
-	// TODO: Revisit during machine remediation implementation which may need to cover other machine phases.
+	return nil, nil
+}
+
+func preflightChecks(controlPlane *internal.ControlPlane) *ctrl.Result {
```
[the remark this answers is not in the document]
Comment (benmoss) [reply controlplane/kubeadm/controllers/controller.go:410]: Yeah, I think all of the scenarios that we want to return early from are just yes/no so we probably can get away with a bool. Name sounds good too.

File: controlplane/kubeadm/controllers/controller.go (hunk 1 above)
Comment (benmoss) [reply controlplane/kubeadm/controllers/controller.go:240]: yeah I tend to abuse the multiline if assignments. Probably just using a regular assignment would be less weird looking

File: controlplane/kubeadm/controllers/controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (benmoss) [reply controlplane/kubeadm/controllers/controller.go:386]: Oh but I also wanted to be able to return the result of `scaleDownControlPlane()`, which is `(ctrl.Result, error)`. Could probably do something like turning the result into a RequeueError but then that just feels gross.

**Comment by benmoss** [comment]
Hard to review now since it contains changes from #2975 but I've done a lot of QA testing with it and it seems to behave how I think it should and recovers from scenarios we want.

- 1 and 2 node clusters do not do any remediation
- 3 node clusters reconcile one machine, but would fail when two machines fail since etcd loses quorum
- Unhealthy machines are removed and replacements are added 1 by 1
- Tested a 9 node cluster recovering from losing 4 nodes
- Upgrades still flow in the same order

**Comment by benmoss** [comment]
One thing I sorta wanted to do on this but didn't was move this [machine phases](https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/machine_controller_phases.go#L46-L84) out into methods on the clusterv1.Machine object. I asked on the Slack if there was some rationale behind not putting methods on API objects, but there doesn't seem to be. Right now this logic is being duplicated.

**Comment by vincepri** [comment]
@benmoss I might have missed the slack conversation, can we file an issue?

**Comment by benmoss** [comment]
> @benmoss I might have missed the slack conversation, can we file an issue?

https://github.com/kubernetes-sigs/cluster-api/issues/3046

**Comment by vincepri** [comment]
Did you want to keep this one open or can we close it?

**Comment by benmoss** [comment]
I'll reopen another one

**Pull request closed without merging**